hs_composite_resource/models.py

```diff
@@ -0,0 +1,180 @@
+                                                  resource_file__contains=path_to_check).all()
+            for res_file_obj in res_file_objs:
+                if not res_file_obj.logical_file.supports_resource_file_rename or \
+                        not res_file_obj.logical_file.supports_resource_file_move:
+                    return False
+
+        return True
+
+    def supports_move_or_rename_file_or_folder(self, src_full_path, tgt_full_path):
+        """checks if file/folder rename/move is allowed"""
+
+        istorage = self.get_irods_storage()
+        tgt_file_dir = os.path.dirname(tgt_full_path)
+        src_file_dir = os.path.dirname(src_full_path)
+
+        def check_target_directory():
+            path_to_check = ''
+            if istorage.exists(tgt_file_dir):
+                path_to_check = tgt_file_dir
+            else:
+                if tgt_file_dir.startswith(src_file_dir):
+                    path_to_check = src_file_dir
+
+            if path_to_check and not path_to_check.endswith("data/contents"):
+                # it is not the base directory - it must be a directory under base dir
+                if self.resource_federation_path:
+                    res_file_objs = self.files.filter(
+                        object_id=self.id,
+                        fed_resource_file_name_or_path__contains=path_to_check).all()
+                else:
+                    res_file_objs = self.files.filter(
+                        object_id=self.id,
+                        resource_file__contains=path_to_check).all()
+                for res_file_obj in res_file_objs:
+                    if not res_file_obj.logical_file.supports_resource_file_rename or \
+                            not res_file_obj.logical_file.supports_resource_file_move:
+                        return False
+            return True
+
+        if self.resource_federation_path:
+            res_file_obj = self.files.filter(
+                object_id=self.id,
+                fed_resource_file_name_or_path=src_full_path).first()
+        else:
+            res_file_obj = self.files.filter(object_id=self.id,
+                                             resource_file=src_full_path).first()
+        if res_file_obj is not None:
+            # src_full_path contains file name
+            if not res_file_obj.logical_file.supports_resource_file_rename or \
+                    not res_file_obj.logical_file.supports_resource_file_move:
+                return False
+
+            # check if the target directory allows stuff to be moved there
+            return check_target_directory()
+        else:
+            # src_full_path is a folder path without file name
+            # tgt_full_path also must be a folder path without file name
+            # check that if the target folder contains any files and if any of those files
+            # allow moving stuff there
+            return check_target_directory()
```

Comment:
The purpose of the function ```check_target_directory()``` to factor out duplicate code. This inner function is needed only within this function. I don't see any reason for making it a utility function. I can rename the function to ```check_directory``` if that helps.